tracer/src/Datadog.Trace/Debugger/Symbols/SymbolBatchUploadApi.cs

```diff
@@ -0,0 +1,133 @@
+                count = Encoding.UTF8.GetBytes(@event, 0, @event.Length, eventAsBytes, 0);
+                if (count == 0)
+                {
+                    Log.Error("Failed to initialize event {Event}", @event);
+                    return default;
+                }
+
+                return new ArraySegment<byte>(eventAsBytes);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e, "Failed to initialize event {Event}", @event);
+                return default;
+            }
+        }
+
+        public static SymbolBatchUploadApi Create(IApiRequestFactory apiRequestFactory, IDiscoveryService discoveryService)
+        {
+            return new SymbolBatchUploadApi(apiRequestFactory, discoveryService);
+        }
+
+        public async Task<bool> SendBatchAsync(ArraySegment<byte> symbols)
+        {
+            if (Volatile.Read(ref _endpoint) is not { } endpoint)
+            {
+                Log.Warning("Failed to upload symbol: symbol endpoint not yet retrieved from discovery service");
+                return false;
+            }
+
+            if (_event == default || _event.Count == 0)
+            {
+                Log.Error("Failed to initialize event");
+                return default;
+            }
+
+            var uri = _apiRequestFactory.GetEndpoint(endpoint);
+            var request = _apiRequestFactory.Create(uri);
+            var multipartRequest = (IMultipartApiRequest)request;
+
+            const int maxRetries = 3;
+            int retries = 0;
+
+            while (retries < maxRetries)
+            {
+                using var response = await multipartRequest.PostAsync(
+                                         new MultipartFormItem("symbols", MimeTypes.Json, null, symbols),
+                                         new MultipartFormItem("event", MimeTypes.Json, "event", _event)).
+                                                            ConfigureAwait(false);
+
+                if (response.StatusCode is >= 200 and <= 299)
+                {
+                    return true;
+                }
+
+                retries++;
+                var content = await response.ReadAsStringAsync().ConfigureAwait(false);
+
+                if (ShouldRetry(response.StatusCode))
+                {
+                    await Task.Delay(GetDelayTime(retries)).ConfigureAwait(false);
+                }
+                else
+                {
+                    Log.Error<int, string>("Failed to upload symbol with status code {StatusCode} and message: {ResponseContent}", response.StatusCode, content);
+                    return false;
+                }
+            }
+
+            return false;
+        }
+
+        private static bool ShouldRetry(int statusCode)
+        {
+            int[] statusCodesToRetry = { 408, 425, 429, 503, 504 };
+            return statusCodesToRetry.Any(code => code == statusCode);
+        }
```

Comment:
Curious, where does the decision to retry for these status codes come from? How come we don't do it anywhere else in our code base?

Reply:
Oh, just found https://github.com/DataDog/dd-trace-dotnet/blob/77a8cb82e6b68e5e5ba877b8a0c34c281496a354/tracer/src/Datadog.Trace/Logging/DirectSubmission/Sink/LogsApi.cs#L92. Should we try to extract and reuse?